File: src/foreground_tracker.py
"""Foreground time tracker for time-based triggers."""
import time


class ForegroundTracker:
    def __init__(self):
        self._active = {}  # trigger_name -> accumulated_seconds
        self._last_update = {}  # trigger_name -> last_seen_timestamp

    def update_active_matches(self, matched_names: list):
        now = time.time()
        current = set(matched_names)

        for name in current:
            if name in self._last_update:
                delta = now - self._last_update[name]
                if delta < 15:  # Only count if seen recently (within tick interval)
                    self._active[name] = self._active.get(name, 0) + delta
            self._last_update[name] = now

        # Clear triggers that are no longer active
        for name in list(self._last_update):
            if name not in current:
                del self._last_update[name]

    def has_exceeded_limit(self, trigger) -> bool:
        if not trigger.is_time_based:
            return False
        accumulated = self._active.get(trigger.name, 0)
        return accumulated >= trigger.time_limit_minutes * 60

    def reset_trigger(self, name: str):
        self._active.pop(name, None)
        self._last_update.pop(name, None)

    def reset_all(self):
        self._active.clear()
        self._last_update.clear()

File: Android/app/src/main/kotlin/com/stickyalarm/domain/ForegroundTracker.kt
package com.stickyalarm.domain

import com.stickyalarm.data.model.TriggerEntry

class ForegroundTracker {
    private val active = mutableMapOf<String, Double>()
    private val lastUpdate = mutableMapOf<String, Double>()
    private val lastSeen = mutableMapOf<String, Long>()

    companion object {
        private const val COOLDOWN_MS = 5 * 60 * 1000L // 5 min cooldown
    }

    fun updateActiveMatches(matchedNames: List<String>) {
        val now = System.currentTimeMillis() / 1000.0
        val nowMs = System.currentTimeMillis()
        val current = matchedNames.toSet()

        for (name in current) {
            val last = lastUpdate[name]
            if (last != null) {
                val delta = now - last
                if (delta < 15) {
                    active[name] = (active[name] ?: 0.0) + delta
                }
            }
            lastUpdate[name] = now
            lastSeen[name] = nowMs
        }

        // Inactive triggers: check cooldown
        for (name in active.keys.toList()) {
            if (name !in current) {
                lastUpdate.remove(name)
                val ls = lastSeen[name] ?: 0L
                if (nowMs - ls >= COOLDOWN_MS) {
                    // 5 min inactive -> reset counter
                    active.remove(name)
                    lastSeen.remove(name)
                }
                // Otherwise: counter stays (no reset on brief pause)
            }
        }
    }

    fun hasExceededLimit(trigger: TriggerEntry): Boolean {
        if (!trigger.isTimeBased) return false
        val accumulated = active[trigger.name] ?: 0.0
        return accumulated >= trigger.timeLimitMinutes * 60
    }

    fun resetTrigger(name: String) {
        active.remove(name)
        lastUpdate.remove(name)
        lastSeen.remove(name)
    }

    fun resetAll() {
        active.clear()
        lastUpdate.clear()
        lastSeen.clear()
    }
}
